### PR TITLE
Switch test_utils to tempfile

### PR DESCRIPTION
**Pull request by sfc-gh-jsikorski**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [ ] I've added new automated tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [x] I've described my changes in the section below.

### Changes description
Change from manually creating directories and files for unit tests for utils.py to files generated by tempfile
+ minor refactor

### CODE REVIEW

File: tests/test_utils.py [hunk 1]
```diff
@@ -456,7 +449,9 @@ def include_paths_env_variable(self, other_directory: str) -> Generator:
         os.environ.pop("SNOWCLI_INCLUDE_PATHS")
 
     @staticmethod
-    def create_file(filepath: str, contents: List[str]) -> None:
-        with open(filepath, "w") as new_file:
-            for line in contents:
-                new_file.write(line + "\n")
+    def create_file(suffix: str, dir: str, contents: List[str]) -> str:
```
Comment (sfc-gh-pjob) [inline tests/test_utils.py:452]:
```suggestion
    def create_temp_file(suffix: str, dir: str, contents: List[str]) -> str:
```

File: tests/test_utils.py [hunk 2]
```diff
@@ -388,66 +399,48 @@ def test_deduplicate_and_sort_reqs(self):
...
-        ).lower()
-        os.mkdir(path)
-        yield path
-        rmtree(path)
+        other_dir = tempfile.TemporaryDirectory()
+        yield other_dir.name
+        rmtree(other_dir.name)
 
     @pytest.fixture
     def file_in_other_directory(self, other_directory: str) -> Generator:
```
Comment (sfc-gh-pjob) [inline tests/test_utils.py:442]:
```suggestion
    def temp_file_in_other_directory(self, other_directory: str) -> Generator:
```

File: tests/test_utils.py [hunk 3]
```diff
@@ -388,66 +399,48 @@ def test_deduplicate_and_sort_reqs(self):
...
-        current_path = Path(os.getcwd())
-        path = os.path.join(
-            current_path.parent.absolute(), self.SECOND_TEST_DIRECTORY
-        ).lower()
-        os.mkdir(path)
-        yield path
-        rmtree(path)
+        other_dir = tempfile.TemporaryDirectory()
+        yield other_dir.name
+        rmtree(other_dir.name)
```
Comment (sfc-gh-pjob) [inline tests/test_utils.py:439]: Do we have to remove directory on our own? Can't we do sth like the following?:

```
with TemporaryDirectory() as tmp_dir:
        yield tmp_dir
```

File: tests/test_utils.py [hunk 4]
```diff
@@ -456,7 +449,9 @@ def include_paths_env_variable(self, other_directory: str) -> Generator:
...
-    def create_file(filepath: str, contents: List[str]) -> None:
-        with open(filepath, "w") as new_file:
-            for line in contents:
-                new_file.write(line + "\n")
+    def create_file(suffix: str, dir: str, contents: List[str]) -> str:
+        with tempfile.NamedTemporaryFile(suffix=suffix, dir=dir, delete=False) as tmp:
+            with open(tmp.name, "w") as new_file:
+                for line in contents:
+                    new_file.write(line + "\n")
+        return tmp.name
```
Comment (sfc-gh-pjob) [inline tests/test_utils.py:457]: Does it return a path or only a name?

If it's only a name, then you won't know the full path when using functions like `file_in_a_subdir` which calls `create_file`.

File: tests/test_utils.py [hunk 5]
```diff
@@ -388,66 +399,48 @@ def test_deduplicate_and_sort_reqs(self):
...
     def correct_metadata_file(self, temp_test_directory: str) -> Generator:
-        path = os.path.join(temp_test_directory, self.CORRECT_METADATA)
-        self.create_file(path, test_data.correct_package_metadata)
-        yield path
+        yield self.create_file(
+            ".yaml", temp_test_directory, test_data.correct_package_metadata
+        )
 
     @pytest.fixture
     def file_in_a_subdir(self, temp_test_directory: str) -> Generator:
```
Comment (sfc-gh-pjob) [inline tests/test_utils.py:431]:
```suggestion
    def txt_file_in_a_subdir(self, temp_test_directory: str) -> Generator:
```

File: tests/test_utils.py [hunk 6]
```diff
@@ -388,66 +399,48 @@ def test_deduplicate_and_sort_reqs(self):
...
-        self.create_file(path, test_data.requirements)
-        yield path
+        yield self.create_file(".txt", temp_test_directory, test_data.requirements)
 
     @pytest.fixture
     def streamlit_requirements_txt(self, temp_test_directory: str) -> Generator:
-        path = os.path.join(temp_test_directory, self.REQUIREMENTS_SNOWFLAKE)
-        self.create_file(path, test_data.requirements)
-        yield path
+        yield self.create_file(".txt", temp_test_directory, test_data.requirements)
```
Comment (sfc-gh-pjob) [inline tests/test_utils.py:422]: It looks the same as `correct_requirements_txt`

File: tests/test_utils.py [hunk 7]
```diff
@@ -388,66 +399,48 @@ def test_deduplicate_and_sort_reqs(self):
...
-        yield path
+        temp_dir = tempfile.TemporaryDirectory(dir=temp_test_directory)
+        yield temp_dir.name
 
     @pytest.fixture
     def correct_app_zip(self, temp_test_directory: str) -> Generator:
-        path = os.path.join(temp_test_directory, self.APP_ZIP)
-        self.create_file(path, [])
-        yield path
+        yield self.create_file(".zip", temp_test_directory, [])
```
Comment (sfc-gh-pjob) [inline tests/test_utils.py:414]: I wonder if `correct_app_zip` is a good name for function creating an empty file only with `.zip` suffix.

File: tests/test_utils.py [hunk 8]
```diff
@@ -388,66 +399,48 @@ def test_deduplicate_and_sort_reqs(self):
 
     @pytest.fixture
     def temp_test_directory(self) -> Generator:
-        path = os.path.join(os.getcwd(), self.TEMP_TEST_DIRECTORY)
-        os.mkdir(path)
-        yield path
-        rmtree(self.TEMP_TEST_DIRECTORY)  # We delete whole directory in teardown -
+        temp_dir = tempfile.TemporaryDirectory()
+        yield temp_dir.name
+        rmtree(temp_dir.name)  # We delete whole directory in teardown -
```
Comment (sfc-gh-pjob) [inline tests/test_utils.py:404]: The same comment about directory removing as to `other_directory` function.

File: tests/test_utils.py [hunk 9]
```diff
@@ -274,20 +281,24 @@ def test_standard_zip_dir_with_env_variable(
...
         assert result == test_data.requirements
 
-    def test_get_snowflake_packages_delta(self, streamlit_requirements_txt):
+    def test_get_snowflake_packages_delta(
+        self, temp_test_directory: str, streamlit_requirements_txt
+    ):
         anaconda_package = test_data.requirements[-1]
-        os.chdir(".tests")
+        os.chdir(temp_test_directory)
+        os.rename(streamlit_requirements_txt, self.REQUIREMENTS_SNOWFLAKE)
```
Comment (sfc-gh-pjob) [inline tests/test_utils.py:300]: Why do we firstly create file with random name to change it later to a specific name?

File: tests/test_utils.py [hunk 10]
```diff
@@ -274,20 +281,24 @@ def test_standard_zip_dir_with_env_variable(
...
 
         assert result == test_data.requirements
 
-    def test_get_snowflake_packages_delta(self, streamlit_requirements_txt):
+    def test_get_snowflake_packages_delta(
+        self, temp_test_directory: str, streamlit_requirements_txt
+    ):
         anaconda_package = test_data.requirements[-1]
-        os.chdir(".tests")
+        os.chdir(temp_test_directory)
```
Comment (sfc-gh-pjob) [inline tests/test_utils.py:299]: Is it safe to stay in a temp dir after a test?

The same question to other usages of `chdir`.

File: tests/test_utils.py [hunk 11]
```diff
@@ -274,20 +281,24 @@ def test_standard_zip_dir_with_env_variable(
             os.path.join("subdir", os.path.basename(file_in_a_subdir))
             not in zip_file.namelist()
         )
-        assert os.path.join(self.FILE_IN_SECOND_TEST_DIRECTORY) in zip_file.namelist()
+        assert Path(file_in_other_directory).name in zip_file.namelist()
 
-    def test_get_snowflake_packages(self, streamlit_requirements_txt):
-        os.chdir(".tests")
+    def test_get_snowflake_packages(
```
Comment (sfc-gh-pjob) [inline tests/test_utils.py:286]: The same 2 comments as to `test_get_snowflake_packages_delta`

File: tests/test_utils.py (hunk 4 above)
Comment (sfc-gh-jsikorski) [reply tests/test_utils.py:457]: Contrary to Path.name (that returns only filename) tempfile .name returns an absolute path

File: tests/test_utils.py (hunk 9 above)
Comment (sfc-gh-pjob) [reply tests/test_utils.py]: If you need a file with a specific name and if you create it in temp directory (what takes care of cleaning), then IMO it's better to create a file with the target name right up front.

File: tests/test_utils.py (hunk 10 above)
Comment (sfc-gh-pjob) [reply tests/test_utils.py:294]: Looks good, but I would wrap it in a fixture.

I have such util in my PR with integration tests (https://github.com/Snowflake-Labs/snowcli/pull/245/files#diff-9ef616b95f50c54f5ca3b20f74dab7bdb579c6abd97a1c9cb8d624449bb6da8dR20).

I don't know if we want to share utils between tests and integration tests, so if you want to, you may simply copy-paste it. WDYT?

```python
class WorkingDirectoryChanger:
    def __init__(self):
        self._initial_working_directory = os.getcwd()

    @staticmethod
    def change_working_directory_to(directory: str):
        os.chdir(directory)

    def restore_initial_working_directory(self):
        self.change_working_directory_to(self._initial_working_directory)


@pytest.fixture
def temporary_working_directory():
    working_directory_changer = WorkingDirectoryChanger()
    with TemporaryDirectory() as tmp_dir:
        working_directory_changer.change_working_directory_to(tmp_dir)
        yield tmp_dir
        working_directory_changer.restore_initial_working_directory()

```

Comment (sfc-gh-jsikorski) [reply tests/test_utils.py]: I tried few different approaches (yours, pure fixture, decorator), and i chose using separate fixture. It`s not as versatile as your Changer, but i only need some tests to enter temp directory, and this way i avoid duplicating large chunks of code

**Review by sfc-gh-pjob — APPROVED** [review]

**Pull request merged**